### ISSUE TITLE
CH4/OFI: (Ring Index out of range) in MPIDI_OFI_handle_cq_entries from MPI_Win_unlock on BGQ

### ISSUE THREAD

**Issue by pkcoff** [pmodels/mpich#2689]
The PlasticNeocortex Mira INCITE project has been evaluating MPICH3-BGQ-OFI mainly because of memory constraints that disallow usage of two-phase collective IO in the bgq driver, but the new one-sided algorithm in MPICH3 does not increase memory footprint so they can actually do collective IO running at 32 ppn.  However, there is a bug where if enough ranks are putting data to a window in the memory on the same proc it fails with this message from the target proc in stderr:

```
192: INTERNAL ERROR: invalid error code dfffada0 (Ring Index out of range) in MPIDI_OFI_handle_cq_entries:792
192: Abort(606232079) on node 192 (rank 192 in comm 0): Fatal error in MPI_Win_unlock: Other MPI error, error stack:
192: MPI_Win_unlock(116).............: MPI_Win_unlock(rank=192, win=0xa0000000) failed
192: MPID_Win_unlock(168)............:
192: MPIDI_NM_mpi_win_unlock(672)....:
192: MPID_Progress_test(44)..........:
192: MPIDI_OFI_handle_cq_entries(792):
```

I have a simple program using `MPI_File_write_all` to reproduce, basically in MPI-IO if the data chunks per proc are small so that it takes a lot of procs to fill the same collective buffer on target aggregator rank (my smallest reproducer is 4096 ranks worth of data (running on 64 nodes c64) will fit in just a couple collective buffers)  it will fail with that message. The failure messages come from a rank with the collective buffer with target window on it. The passive one-sided interaction in ROMIO for this case is very straightforward, just:

```
MPI_Win_lock (collective buffer on rank 192)
MPI_Put (1 contiguous chunk of data)
MPI_Win_unlock (collective buffer on rank 192)
```

So say the window is on the collective buffer on rank 192 and a couple thousand ranks  are all doing the above to put a few bytes of data into it the test program fails with the above message.   Codewise it crashes on line 792 in `ofi_events.h` on the `MPIDI_OFI_dispatch_function` call:

```
MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_handle_cq_entries(struct fi_cq_tagged_entry *wc,
                                                         ssize_t num, int buffered)
{
    int i, mpi_errno = MPI_SUCCESS;
    MPIR_Request *req;
    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_NETMOD_HANDLE_CQ_ENTRIES);
    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_NETMOD_HANDLE_CQ_ENTRIES);

    for (i = 0; i < num; i++) {
        req = MPIDI_OFI_context_to_request(wc[i].op_context);
        MPIDI_OFI_MPI_CALL_POP(MPIDI_OFI_dispatch_function(&wc[i], req, buffered));
    }

  fn_exit:
    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_NETMOD_HANDLE_CQ_ENTRIES);
    return mpi_errno;
  fn_fail:
    goto fn_exit;
}
```

I guess I'm looking for suggestions on what could be overflowing with this completion queue processing, or what other underlying problem this message is indicating, is there some limit somewhere that i'm somehow exceeding?

**Comment by pkcoff** [comment]
A couple more notes:
1.)  So at 64 ppn I can get that error on just 64 nodes, however at 32ppn I have to run on 2048 nodes and then instead of the error I get a hang in the unlock.
2.) I tried writing a simpler reproducer program which just did the one-sided stuff and it succeeds, so has something to do with the state of stack / one-sided algorithm.
3.) In the process of trying to debug this I put some print statements in and that changes the characteristics of the failure (may fail in the lock instead of unlock, only need 32 nodes at 64ppn to get the failure, etc) so seems some sort of corruption may be going on.

I've spent quite a bit of time going thru the code, on the bgq provider side these queues are implemented as linked lists which should be able to pretty much grow indefinitely until we run out of heap, may be best to meet with Ken and Mike on Monday to walk thru some code to verify it all makes sense i guess.....

**Comment by pkcoff** [comment]
Oh i should also mention this in manual progress mode.  For auto progress mode, i can run only 16 ppn and it only fails in the cases where I add print trace statements to interupt the timing / stack, with no print statements it seems to always work in autoprogress mode....

**Comment by mblockso** [comment]
FYI @hajimefu

**Comment by pkcoff** [comment]
Met with Mike, Ken and Hajime --- most likely issue with multi_recv on mpich or ofi sides, I have todos to gather more info and will be having working sessions with Mike and Hajime hopefully thoughout this week until resolution

**Comment by pkcoff** [comment]
Resolved, PRs for both mpich:
https://github.com/pmodels/mpich/pull/2702
and OFI:
https://github.com/ofiwg/libfabric/pull/3190
Thanks to Mike and Hajime for their consultation - closing.

**Comment by wesbland** [comment]
Maybe we should wait to close this until the PRs are merged? Usually this gets handled by putting

> Closes pmodels/mpich#2702

in the commit message.
